File: web/src/mocks/handlers/seriesExports.ts
/**
 * MSW handlers for series export API endpoints
 */

import { delay, HttpResponse, http } from "msw";
import type { components } from "@/types/api.generated";

type SeriesExportDto = components["schemas"]["SeriesExportDto"];
type SeriesExportListResponse =
  components["schemas"]["SeriesExportListResponse"];
type ExportFieldDto = components["schemas"]["ExportFieldDto"];
type ExportFieldCatalogResponse =
  components["schemas"]["ExportFieldCatalogResponse"];

// Mock data store
let mockExports: SeriesExportDto[] = [
  {
    id: "550e8400-e29b-41d4-a716-446655440001",
    format: "json",
    status: "completed",
    libraryIds: ["lib-1"],
    fields: ["title", "summary", "genres", "user_rating"],
    fileSizeBytes: 24576,
    rowCount: 42,
    error: null,
    createdAt: new Date(Date.now() - 86400000).toISOString(), // 1 day ago
    startedAt: new Date(Date.now() - 86400000 + 1000).toISOString(),
    completedAt: new Date(Date.now() - 86400000 + 5000).toISOString(),
    expiresAt: new Date(Date.now() + 6 * 86400000).toISOString(), // 6 days from now
  },
  {
    id: "550e8400-e29b-41d4-a716-446655440002",
    format: "csv",
    status: "running",
    libraryIds: ["lib-1", "lib-2"],
    fields: [
      "title",
      "summary",
      "genres",
      "tags",
      "authors",
      "actual_book_count",
    ],
    fileSizeBytes: null,
    rowCount: null,
    error: null,
    createdAt: new Date(Date.now() - 60000).toISOString(), // 1 min ago
    startedAt: new Date(Date.now() - 55000).toISOString(),
    completedAt: null,
    expiresAt: new Date(Date.now() + 7 * 86400000).toISOString(),
  },
  {
    id: "550e8400-e29b-41d4-a716-446655440003",
    format: "json",
    status: "failed",
    libraryIds: ["lib-3"],
    fields: ["title", "genres"],
    fileSizeBytes: null,
    rowCount: null,
    error: "disk full",
    createdAt: new Date(Date.now() - 3600000).toISOString(), // 1 hour ago
    startedAt: new Date(Date.now() - 3599000).toISOString(),
    completedAt: new Date(Date.now() - 3590000).toISOString(),
    expiresAt: new Date(Date.now() + 7 * 86400000).toISOString(),
  },
];

const mockFieldCatalog: ExportFieldDto[] = [
  {
    key: "series_id",
    label: "Series ID",
    multiValue: false,
    userSpecific: false,
  },
  {
    key: "series_name",
    label: "Series Name",
    multiValue: false,
    userSpecific: false,
  },
  {
    key: "library_id",
    label: "Library ID",
    multiValue: false,
    userSpecific: false,
  },
  {
    key: "library_name",
    label: "Library Name",
    multiValue: false,
    userSpecific: false,
  },
  { key: "path", label: "Path", multiValue: false, userSpecific: false },
  {
    key: "created_at",
    label: "Created At",
    multiValue: false,
    userSpecific: false,
  },
  {
    key: "updated_at",
    label: "Updated At",
    multiValue: false,
    userSpecific: false,
  },
  { key: "title", label: "Title", multiValue: false, userSpecific: false },
  { key: "summary", label: "Summary", multiValue: false, userSpecific: false },
  {
    key: "publisher",
    label: "Publisher",
    multiValue: false,
    userSpecific: false,
  },
  { key: "status", label: "Status", multiValue: false, userSpecific: false },
  { key: "year", label: "Year", multiValue: false, userSpecific: false },
  {
    key: "language",
    label: "Language",
    multiValue: false,
    userSpecific: false,
  },
  { key: "authors", label: "Authors", multiValue: true, userSpecific: false },
  { key: "genres", label: "Genres", multiValue: true, userSpecific: false },
  { key: "tags", label: "Tags", multiValue: true, userSpecific: false },
  {
    key: "alternate_titles",
    label: "Alternate Titles",
    multiValue: true,
    userSpecific: false,
  },
  {
    key: "expected_book_count",
    label: "Expected Book Count",
    multiValue: false,
    userSpecific: false,
  },
  {
    key: "actual_book_count",
    label: "Actual Book Count",
    multiValue: false,
    userSpecific: false,
  },
  {
    key: "unread_book_count",
    label: "Unread Book Count",
    multiValue: false,
    userSpecific: false,
  },
  {
    key: "user_rating",
    label: "User Rating",
    multiValue: false,
    userSpecific: true,
  },
  {
    key: "user_notes",
    label: "User Notes",
    multiValue: false,
    userSpecific: true,
  },
  {
    key: "community_avg_rating",
    label: "Community Avg Rating",
    multiValue: false,
    userSpecific: false,
  },
  {
    key: "external_ratings",
    label: "External Ratings",
    multiValue: true,
    userSpecific: false,
  },
];

export const seriesExportsHandlers = [
  // List exports
  http.get("/api/v1/user/exports/series", async () => {
    await delay(150);
    const response: SeriesExportListResponse = { exports: mockExports };
    return HttpResponse.json(response);
  }),

  // Get field catalog
  http.get("/api/v1/user/exports/series/fields", async () => {
    await delay(100);
    const response: ExportFieldCatalogResponse = { fields: mockFieldCatalog };
    return HttpResponse.json(response);
  }),

  // Get single export
  http.get("/api/v1/user/exports/series/:id", async ({ params }) => {
    await delay(100);
    const exp = mockExports.find((e) => e.id === params.id);
    if (!exp) {
      return HttpResponse.json({ error: "Export not found" }, { status: 404 });
    }
    return HttpResponse.json(exp);
  }),

  // Create export
  http.post("/api/v1/user/exports/series", async ({ request }) => {
    await delay(300);
    const body = (await request.json()) as {
      format: string;
      libraryIds: string[];
      fields: string[];
    };

    const newExport: SeriesExportDto = {
      id: crypto.randomUUID(),
      format: body.format,
      status: "pending",
      libraryIds: body.libraryIds,
      fields: body.fields,
      fileSizeBytes: null,
      rowCount: null,
      error: null,
      createdAt: new Date().toISOString(),
      startedAt: null,
      completedAt: null,
      expiresAt: new Date(Date.now() + 7 * 86400000).toISOString(),
    };

    mockExports = [newExport, ...mockExports];

    // Simulate task completing after 3 seconds
    setTimeout(() => {
      const idx = mockExports.findIndex((e) => e.id === newExport.id);
      if (idx >= 0) {
        mockExports[idx] = {
          ...mockExports[idx],
          status: "completed",
          startedAt: new Date().toISOString(),
          completedAt: new Date().toISOString(),
          rowCount: Math.floor(Math.random() * 100) + 5,
          fileSizeBytes: Math.floor(Math.random() * 100000) + 1000,
        };
      }
    }, 3000);

    return HttpResponse.json(newExport, { status: 202 });
  }),

  // Delete export
  http.delete("/api/v1/user/exports/series/:id", async ({ params }) => {
    await delay(200);
    const idx = mockExports.findIndex((e) => e.id === params.id);
    if (idx < 0) {
      return HttpResponse.json({ error: "Export not found" }, { status: 404 });
    }
    mockExports = mockExports.filter((e) => e.id !== params.id);
    return new HttpResponse(null, { status: 204 });
  }),

  // Download export (returns mock blob)
  http.get("/api/v1/user/exports/series/:id/download", async ({ params }) => {
    await delay(200);
    const exp = mockExports.find((e) => e.id === params.id);
    if (!exp || exp.status !== "completed") {
      return HttpResponse.json({ error: "Export not ready" }, { status: 409 });
    }

    const mockContent =
      exp.format === "json"
        ? JSON.stringify(
            [
              {
                series_id: "1",
                series_name: "Mock Series A",
                library_id: "lib-1",
              },
              {
                series_id: "2",
                series_name: "Mock Series B",
                library_id: "lib-1",
              },
            ],
            null,
            2,
          )
        : "series_id,series_name,library_id\n1,Mock Series A,lib-1\n2,Mock Series B,lib-1\n";

    const contentType = exp.format === "csv" ? "text/csv" : "application/json";

    return new HttpResponse(mockContent, {
      headers: {
        "Content-Type": contentType,
        "Content-Disposition": `attachment; filename="codex-export.${exp.format}"`,
      },
    });
  }),
];

File: src/api/routes/v1/handlers/series_exports.rs
//! Handlers for series export endpoints

use axum::{
    Json,
    extract::{Path, State},
    http::{StatusCode, header},
    response::Response,
};
use chrono::{Duration, Utc};
use std::sync::Arc;
use uuid::Uuid;

use crate::api::error::ApiError;
use crate::api::extractors::auth::{AppState, AuthContext};
use crate::db::repositories::{SeriesExportRepository, TaskRepository};
use crate::services::export_storage::{DEFAULT_EXPORTS_DIR, ExportStorage};
use crate::services::series_export_collector::ExportField;
use crate::tasks::types::TaskType;

use super::super::dto::series_export::{
    CreateSeriesExportRequest, ExportFieldCatalogResponse, ExportFieldDto, SeriesExportDto,
    SeriesExportListResponse,
};

/// Default concurrent export limit per user
const DEFAULT_MAX_CONCURRENT: u64 = 3;
/// Default retention days
const DEFAULT_RETENTION_DAYS: u64 = 7;

/// POST /user/exports/series - Create a new series export job
pub async fn create_export(
    State(state): State<Arc<AppState>>,
    auth: AuthContext,
    Json(request): Json<CreateSeriesExportRequest>,
) -> Result<(StatusCode, Json<SeriesExportDto>), ApiError> {
    let user_id = auth.user_id;

    // Validate format
    if request.format != "json" && request.format != "csv" {
        return Err(ApiError::BadRequest(
            "Format must be 'json' or 'csv'".to_string(),
        ));
    }

    // Validate at least one library
    if request.library_ids.is_empty() {
        return Err(ApiError::BadRequest(
            "At least one library must be selected".to_string(),
        ));
    }

    // Validate fields - must all be valid field keys
    for key in &request.fields {
        if ExportField::parse(key).is_none() {
            return Err(ApiError::BadRequest(format!("Unknown field: {key}")));
        }
    }

    // Check concurrent export limit
    let settings = &state.settings_service;
    let max_concurrent = settings
        .get_uint("exports.max_concurrent_per_user", DEFAULT_MAX_CONCURRENT)
        .await
        .unwrap_or(DEFAULT_MAX_CONCURRENT);

    let active_count = SeriesExportRepository::count_non_terminal_by_user(&state.db, user_id)
        .await
        .map_err(|e| ApiError::Internal(format!("Failed to check active exports: {e}")))?;

    if active_count >= max_concurrent {
        return Err(ApiError::Conflict(format!(
            "You already have {active_count} active exports (limit: {max_concurrent})"
        )));
    }

    // Compute expiry
    let retention_days = settings
        .get_uint("exports.retention_days", DEFAULT_RETENTION_DAYS)
        .await
        .unwrap_or(DEFAULT_RETENTION_DAYS);
    let expires_at = Utc::now() + Duration::days(retention_days as i64);

    // Create export record
    let library_ids_json = serde_json::to_value(&request.library_ids)
        .map_err(|e| ApiError::Internal(format!("Failed to serialize library_ids: {e}")))?;
    let fields_json = serde_json::to_value(&request.fields)
        .map_err(|e| ApiError::Internal(format!("Failed to serialize fields: {e}")))?;

    let export = SeriesExportRepository::create(
        &state.db,
        user_id,
        &request.format,
        library_ids_json,
        fields_json,
        expires_at,
    )
    .await
    .map_err(|e| ApiError::Internal(format!("Failed to create export: {e}")))?;

    // Enqueue background task
    let task_type = TaskType::ExportSeries {
        export_id: export.id,
        user_id,
    };

    TaskRepository::enqueue(&state.db, task_type, None)
        .await
        .map_err(|e| ApiError::Internal(format!("Failed to enqueue export task: {e}")))?;

    let dto = SeriesExportDto::from_model(&export);
    Ok((StatusCode::ACCEPTED, Json(dto)))
}

/// GET /user/exports/series - List current user's exports
pub async fn list_exports(
    State(state): State<Arc<AppState>>,
    auth: AuthContext,
) -> Result<Json<SeriesExportListResponse>, ApiError> {
    let exports = SeriesExportRepository::list_by_user(&state.db, auth.user_id)
        .await
        .map_err(|e| ApiError::Internal(format!("Failed to list exports: {e}")))?;

    let dtos: Vec<SeriesExportDto> = exports.iter().map(SeriesExportDto::from_model).collect();

    Ok(Json(SeriesExportListResponse { exports: dtos }))
}

/// GET /user/exports/series/{id} - Get a single export's details
pub async fn get_export(
    State(state): State<Arc<AppState>>,
    auth: AuthContext,
    Path(id): Path<Uuid>,
) -> Result<Json<SeriesExportDto>, ApiError> {
    let export = SeriesExportRepository::find_by_id_and_user(&state.db, id, auth.user_id)
        .await
        .map_err(|e| ApiError::Internal(format!("Failed to get export: {e}")))?
        .ok_or_else(|| ApiError::NotFound("Export not found".to_string()))?;

    Ok(Json(SeriesExportDto::from_model(&export)))
}

/// GET /user/exports/series/{id}/download - Download the export file
pub async fn download_export(
    State(state): State<Arc<AppState>>,
    auth: AuthContext,
    Path(id): Path<Uuid>,
) -> Result<Response, ApiError> {
    let export = SeriesExportRepository::find_by_id_and_user(&state.db, id, auth.user_id)
        .await
        .map_err(|e| ApiError::Internal(format!("Failed to get export: {e}")))?
        .ok_or_else(|| ApiError::NotFound("Export not found".to_string()))?;

    if export.status != "completed" {
        return Err(ApiError::Conflict(format!(
            "Export is not ready for download (status: {})",
            export.status
        )));
    }

    let file_path = export.file_path.as_deref().ok_or_else(|| {
        ApiError::Internal("Export completed but file path is missing".to_string())
    })?;

    // Read file contents
    let data = tokio::fs::read(file_path).await.map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            ApiError::NotFound("Export file no longer exists on disk".to_string())
        } else {
            ApiError::Internal(format!("Failed to read export file: {e}"))
        }
    })?;

    let content_type = match export.format.as_str() {
        "csv" => "text/csv; charset=utf-8",
        _ => "application/json; charset=utf-8",
    };

    let ext = match export.format.as_str() {
        "csv" => "csv",
        _ => "json",
    };

    let timestamp = export.created_at.format("%Y%m%d_%H%M%S");
    let filename = format!("codex-series-export-{timestamp}.{ext}");

    Ok(Response::builder()
        .status(StatusCode::OK)
        .header(header::CONTENT_TYPE, content_type)
        .header(header::CONTENT_LENGTH, data.len())
        .header(
            header::CONTENT_DISPOSITION,
            format!("attachment; filename=\"{filename}\""),
        )
        .body(axum::body::Body::from(data))
        .unwrap())
}

/// DELETE /user/exports/series/{id} - Delete an export and its file
pub async fn delete_export(
    State(state): State<Arc<AppState>>,
    auth: AuthContext,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    let export = SeriesExportRepository::find_by_id_and_user(&state.db, id, auth.user_id)
        .await
        .map_err(|e| ApiError::Internal(format!("Failed to get export: {e}")))?
        .ok_or_else(|| ApiError::NotFound("Export not found".to_string()))?;

    // Delete file if it exists
    if let Some(ref file_path) = export.file_path {
        let _ = tokio::fs::remove_file(file_path).await;
    }

    // Also try deleting via ExportStorage if available
    let storage = get_export_storage(&state).await;
    let _ = storage.delete(auth.user_id, id, &export.format).await;

    // Delete DB record
    SeriesExportRepository::delete_by_id(&state.db, id)
        .await
        .map_err(|e| ApiError::Internal(format!("Failed to delete export: {e}")))?;

    Ok(StatusCode::NO_CONTENT)
}

/// GET /user/exports/series/fields - Get the field catalog
pub async fn get_field_catalog() -> Json<ExportFieldCatalogResponse> {
    let fields: Vec<ExportFieldDto> = ExportField::ALL
        .iter()
        .map(|f| ExportFieldDto {
            key: f.as_str().to_string(),
            label: field_label(f),
            multi_value: f.is_multi_value(),
            user_specific: f.is_user_specific(),
        })
        .collect();

    Json(ExportFieldCatalogResponse { fields })
}

/// Human-readable label for each export field
fn field_label(field: &ExportField) -> String {
    match field {
        ExportField::SeriesId => "Series ID",
        ExportField::SeriesName => "Series Name",
        ExportField::LibraryId => "Library ID",
        ExportField::LibraryName => "Library Name",
        ExportField::Path => "Path",
        ExportField::CreatedAt => "Created At",
        ExportField::UpdatedAt => "Updated At",
        ExportField::Title => "Title",
        ExportField::Summary => "Summary",
        ExportField::Publisher => "Publisher",
        ExportField::Status => "Status",
        ExportField::Year => "Year",
        ExportField::Language => "Language",
        ExportField::Authors => "Authors",
        ExportField::Genres => "Genres",
        ExportField::Tags => "Tags",
        ExportField::AlternateTitles => "Alternate Titles",
        ExportField::ExpectedBookCount => "Expected Book Count",
        ExportField::ActualBookCount => "Actual Book Count",
        ExportField::UnreadBookCount => "Unread Book Count",
        ExportField::UserRating => "User Rating",
        ExportField::UserNotes => "User Notes",
        ExportField::CommunityAvgRating => "Community Avg Rating",
        ExportField::ExternalRatings => "External Ratings",
    }
    .to_string()
}

/// Get or create an ExportStorage from settings
async fn get_export_storage(state: &AppState) -> ExportStorage {
    let dir = state
        .settings_service
        .get_string("exports.dir", DEFAULT_EXPORTS_DIR)
        .await
        .unwrap_or_else(|_| DEFAULT_EXPORTS_DIR.to_string());
    ExportStorage::new(dir)
}
